### PR TITLE
Adds xml support in UIRichText

### PR DESCRIPTION
**Pull request by ricardoquesada**
Supported tags:
- < small>: makes the font 20% smaller
- < big>: makes the font 25% bigger
- < img src="" />: to add an image
- < font face="path to font" size="size" color="#00ffaa">: font
  attributes
- < i>: italics
- < b>: bold
- < u>: underline
- < del> strike-through
- < br/>: new line

Also adds support for strikethrough, italics, bold and underlie to
`Label`

Contains tests for UIRichText and Label

Squashed commit of the following:

commit a0d6ae0513d40dbf414930210ab032a49d8984a0
Author: Ricardo Quesada ricardoquesada@gmail.com
Date:   Tue Jan 19 20:39:21 2016 -0800

```
cleanup LabelTestNew
```

commit 0cf423af85d88a42fc1317207feeb3d05da7f8ab
Author: Ricardo Quesada ricardoquesada@gmail.com
Date:   Tue Jan 19 20:34:53 2016 -0800

```
adds maaaany tests

...and fixes width and height in <img src="">
```

commit e8ba6acd5d79bbc766d7aa02ef166e166c801d01
Author: Ricardo Quesada ricardoquesada@gmail.com
Date:   Fri Jan 15 19:41:37 2016 -0800

```
Label + URL working Ok
```

commit 36689e29ee3d77310e92264a09084cc06e30664c
Author: Ricardo Quesada ricardoquesada@gmail.com
Date:   Fri Jan 15 18:24:44 2016 -0800

```
adding this code just in case...

... should use a listener component instead
```

commit af03708950e74483b875d0baad593aa6ed242a04
Author: Ricardo Quesada ricardoquesada@gmail.com
Date:   Thu Jan 14 20:23:36 2016 -0800

```
URL node

plus <de><u> and <a img
```

commit e3a4930b012c0b3756752dac6ffb5ad43d24ae99
Author: Ricardo Quesada ricardoquesada@gmail.com
Date:   Wed Jan 13 18:42:18 2016 -0800

```
strikethrough and underline implemented

they support multiline and horizontal alignment as well
includes test
however vertical alignment might not work
```

commit 640ccf39f56e153db46785a61be7000e898174c7
Author: Ricardo Quesada ricardoquesada@gmail.com
Date:   Tue Jan 12 18:14:57 2016 -0800

```
bug fixes
```

commit 5e41fb76e91c571639585a609a255eb41797a302
Author: Ricardo Quesada ricardoquesada@gmail.com
Date:   Tue Jan 12 16:47:20 2016 -0800

```
adds italics and bold "disable test"
```

commit 202c5a45bb9c8ea160b9f6880ef858874e07814b
Author: Ricardo Quesada ricardoquesada@gmail.com
Date:   Tue Jan 12 14:40:13 2016 -0800

```
adds italics test
```

commit d1a8b421445053cc36860fc376f52692a3672dfe
Author: Ricardo Quesada ricardoquesada@gmail.com
Date:   Mon Jan 11 20:53:23 2016 -0800

```
italics is working ok
```

commit fdd02087fce920c27c2409301ec88685fe68085b
Author: Ricardo Quesada ricardoquesada@gmail.com
Date:   Fri Jan 8 17:18:56 2016 -0800

```
color, size and face working
```

commit c01bdef6b1d49f8805b69d4c162b74cd00c8f5b3
Author: Ricardo Quesada ricardoquesada@gmail.com
Date:   Thu Jan 7 19:02:16 2016 -0800

```
initial commit
```

### CODE REVIEW

**Comment by zilongshanren** [comment]
@ricardoquesada  
:+1:

**Comment by Naruto** [comment]
@ricardoquesada :+1:

**Comment by linshun** [comment]
@ricardoquesada Awesome, please merge it into v3 branch, and looking forward to the JavaScript Binding version.

**Comment by zilongshanren** [comment]
Found tests issues:
- 65: Testing Bold
  When I click the "disable bold" menu item, the bmfont bold label will shift on the screen.

image

**Comment by ricardoquesada** [comment]
@zilongshanren 
thanks. yes, there are actually 3 bugs that I know of:
- Label effects: When using some effects, the labels are the incorrect position. I guess the bug is in the old effects code, and the old tests weren't triggering this bug
- underline and strikethrough don't work when using system fonts (limitation in my implementation... I have to double check it)
- alignment is broken in UIRichText when forcing a manual new line. Triggers old bug in the RichText alignment logic.

**Comment by zilongshanren** [comment]
One suggestion about the label wrap option:

image

The English word should wrap by space and Japanese, Chinese and Korean(aka, CJK) should wrap by character.  In other words, only CJK should wrap by character and other languages should wrap by space.

I know this is a limitation in the old implementation, if you could improve it in this PR, it will be appreciated. :smile:

**Comment by ricardoquesada** [comment]
@zilongshanren sure... I'll try to fix it. Seems like an important bug.

**Comment by zilongshanren** [comment]
@ricardoquesada 

> underline and strikethrough don't work when using system fonts (limitation in my implementation... I have to double check it)

The system font part is handled on each platform and we might need to rewrite the logic on every supported platform including the Overflow options I added in v3.10.

**Comment by ricardoquesada** [comment]
thanks. merged... I'll add the JS tests later and fix the remaining bugs.

**Comment by zilongshanren** [comment]
:+1:  @ricardoquesada  thanks.

File: cocos/2d/CCLabel.cpp [hunk 1]
```diff
@@ -207,7 +211,6 @@ Label* Label::createWithSystemFont(const std::string& text, const std::string& f
         return ret;
     }
 
-    delete ret;
```
Comment (zilongshanren) [inline cocos/2d/CCLabel.cpp:210]: why remove this line here?

File: cocos/ui/UIRichText.cpp [hunk 2]
```diff
@@ -373,12 +801,23 @@ void RichText::handleTextRenderer(const std::string& text, const std::string& fo
     }
 }
     
-void RichText::handleImageRenderer(const std::string& fileParh, const Color3B &color, GLubyte opacity)
+void RichText::handleImageRenderer(const std::string& filePath, const Color3B &color, GLubyte opacity, int width, int height)
 {
-    Sprite* imageRenderer = Sprite::create(fileParh);
+    Sprite* imageRenderer = Sprite::create(filePath);
```
Comment (zilongshanren) [inline cocos/ui/UIRichText.cpp:806]: I suggest to use Scale9Sprite with `RenderingType::Simple` here. The code will be more compact. We don't need to calculate the scale.

**Comment by ricardoquesada** [comment]
@zilongshanren

I'm trying to fix the position bug in the "bold" example, but it is working Ok for me... strange.
BTW, what is a "system font"... and why does it behave differently ?

**Comment by zilongshanren** [comment]
@ricardoquesada 
I tested with cpp-test Mac project, you don't have the problem...wired... 

A system font is rendered with the following process:
1. create a font with System API. eg, NSFont in Mac. Here we could provide many options to config the font property.
2. render the font to a memory image.
3. Draw the image with quads.

Every time we change the "system font", we have to redo the above process and it's very slow.

And what we can do with system font is also limited by the specific platforms.

**Comment by ricardoquesada** [comment]
@zilongshanren thanks. I found the problem with underline/italics in system font... fixing it.

I'll try to reproduce the bold bug... I know it is there because I saw it a few days ago... but not today :(
Most probably an ivar is not initialized... working on it.

**Comment by ricardoquesada** [comment]
@zilongshanren  @linshun 
js tests + bug fixes here: https://github.com/cocos2d/cocos2d-x/pull/14930

**Comment by ricardoquesada** [comment]
and... bold bug fixed.

**Comment by zilongshanren** [comment]
@ricardoquesada 
thanks.

**Pull request merged**